source/common/network/happy_eyeballs_connection_impl.cc

```diff
@@ -0,0 +1,433 @@
+#include "source/common/network/happy_eyeballs_connection_impl.h"
+
+#include <vector>
+
+namespace Envoy {
+namespace Network {
+
+HappyEyeballsConnectionImpl::HappyEyeballsConnectionImpl(
+    Event::Dispatcher& dispatcher, const std::vector<Address::InstanceConstSharedPtr>& address_list,
+    Address::InstanceConstSharedPtr source_address, TransportSocketFactory& socket_factory,
+    TransportSocketOptionsConstSharedPtr transport_socket_options,
+    const ConnectionSocket::OptionsSharedPtr options)
+    : dispatcher_(dispatcher), address_list_(address_list), source_address_(source_address),
+      socket_factory_(socket_factory), transport_socket_options_(transport_socket_options),
+      options_(options),
+      next_attempt_timer_(dispatcher_.createTimer([this]() -> void { tryAnotherConnection(); })) {
+  connections_.push_back(createNextConnection());
+}
+
+HappyEyeballsConnectionImpl::~HappyEyeballsConnectionImpl() = default;
+
+void HappyEyeballsConnectionImpl::connect() {
+  ASSERT(!connect_finished_);
+  connections_[0]->connect();
+  maybeScheduleNextAttempt();
+}
+
+void HappyEyeballsConnectionImpl::addWriteFilter(WriteFilterSharedPtr filter) {
+  ASSERT(connect_finished_);
+  connections_[0]->addWriteFilter(filter);
+}
+
+void HappyEyeballsConnectionImpl::addFilter(FilterSharedPtr filter) {
+  ASSERT(connect_finished_);
+  connections_[0]->addFilter(filter);
+}
+
+void HappyEyeballsConnectionImpl::addReadFilter(ReadFilterSharedPtr filter) {
+  if (connect_finished_) {
+    connections_[0]->addReadFilter(filter);
+    return;
+  }
+  post_connect_state_.read_filters_.push_back(filter);
+}
+
+void HappyEyeballsConnectionImpl::removeReadFilter(ReadFilterSharedPtr filter) {
+  ASSERT(connect_finished_);
+  connections_[0]->removeReadFilter(filter);
```

Comment:
What will happen if remove is called before connect_finished? Do we need to clear it up in post_connect_state_?

Reply:
Right! So the idea is that some of these methods are permitted to be called before connect() completes, and some of them are not. This is one of the methods in the latter category. It is illegal to be called before connect() completes. (That's what the ASSERT(connect_finished_) is all about).

WDYT?

Reply:
Ah huh! I somehow messed up the ASSERT here with the if statement above. Clearly I can't read....

Reply:
May be worth a comment (here or in envoy/...) why add can be called before connect and remove can't

Reply:
Agreed. But this raises a question that I'm not sure about... is it "kosher" for me to do this? I generated the set of ASSERT()s by putting ASSERTS in all the methods and removing those until all the test/integration/... passed. (A bit more setup to get in this state but that's where the list came from).

Alternatively, I could remove the ASSERT (now ENVOY_BUG) from each of these methods and instead allow them to be called before connect finishes.  (All the setters could make use of post_connect_state_. The getters could just return connection_[0]s info). Maybe that is safer. WDYT?

Reply:
your call! I think it's fine to add bugs/asserts to that code and ask folks to keep an eye out in production, or handle all calls happening early.   I just don't want refactors to the connection class which will affect the happy eyeballs cases not getting caught in production :-)

Reply:
After a brief discussion with Matt, I decided to remove these restrictions and permit all the methods to be called at anytime.